mobile_app_connector/models/firebase_notification.py

```diff
@@ -0,0 +1,147 @@
+#    @author: Nicolas Badoux <n.badoux@hotmail.com>
+#
+#    The licence is in the file __manifest__.py
+#
+##############################################################################
+
+from odoo import api, fields, models
+from odoo.fields import Datetime
+from datetime import datetime
+
+
+class FirebaseNotification(models.Model):
+    _inherit = "firebase.notification"
+
+    destination = fields.Selection([
+        ('MyHub', 'MyHub'),
+        ('Letter', 'Letter'),
+        ('Donation', 'Donation'),
+        ('Prayer', 'Prayer'),
+    ], default="MyHub")
+
+    topic = fields.Selection([
+        ('general_notification', 'General channel'),
+        ('child_notification', 'Child channel'),
+        ('spam', "None (bypass user's preferences)")
+    ], default='general_notification', required=True)
+
+    fundType = fields.Many2one('product.product')
+
+    ##########################################################################
+    #                             PUBLIC METHODS                             #
+    ##########################################################################
+
+    @api.multi
+    def send(self, data=None):
+        """
+        Filters notifications w.r.t. user's preference
+
+        :param data:
+        :return:
+        """
+        for notif in self:
+            data = {
+                'topic': notif.topic,
+                "destination": notif.destination or "",
+                "fund_type_id": str(notif.fundType.id)
+            }
+            super(FirebaseNotification, self).send(data)
+
+
+    def notification_cron(self):
+        """
+        Overriding so the automated notifications are filtered to
+        :return:
+        """
+        super(FirebaseNotification, self).notification_cron()
+
+    def duplicate_to_unread(self):
+        res = super(FirebaseNotification, self).duplicate_to_unread()
+        new = self.browse(res['res_id'])
+        new.destination = self.destination
+        new.fundType = self.fundType
+        return res
+
+    @api.model
+    def mobile_get_notification(self, **params):
+        """
+        This is called when the app retrieves the notification.
+        :param params: {
+            firebase_id: the device id requesting its notifications,
+            supid: id of the partner logged in the app
+        }
+        :return: a list of notifications as expected by the app
+        """
+        firebase_id = params.get('firebase_id')
+        reg = self.env['firebase.registration']\
+            .search([('registration_id', '=', firebase_id)])
+        if reg.partner_id:
+            dt = datetime.strftime(datetime.today(), "%Y-%m-%d %H:%M:%S")
+            notifications = self.search([
+                ('partner_ids', '=', reg.partner_id.id),
+                ('send_date', '<', dt),
+                ('sent', '=', True)
+            ])
+        else:
+            # Logged out users
+            dt = datetime.strftime(datetime.today(), "%Y-%m-%d %H:%M:%S")
```

Comment:
Same `dt` as above can be used if the declaration is moved before the `if` clause.